### PR TITLE
Yet another pass at connection cleanup.

### LINKED ISSUE #805: Connection pool queue can still be exhausted by gevent timeouts
A similar error was discussed here https://github.com/shazow/urllib3/issues/644 already and some fix applied, but the problem still partly exists. Consider this code:

```python
from gevent import monkey, spawn, sleep, Timeout, joinall
monkey.patch_all()

import sys
from itertools import product
from urllib3.connectionpool import HTTPConnectionPool

def timed_read(cp, i, timeout, release=None, preload_content=True):
    try:
        with Timeout(timeout):
            resp = cp.urlopen('GET', 'http://httpbin.org/redirect/:1',
                              release_conn=release,
                              preload_content=preload)
            print "%s opened" % i
            sys.stdout.flush()
            data = resp.data
    except Timeout:
        print "%s timed out" % i
    else:
        content_len = int(resp.headers['content-length'])
        if len(data) != content_len:
            print "%s incomplete read: %s/%s" % (i, len(data), content_len)
        else:
            print "%s finished" % i
    sys.stdout.flush()


def main(timeout=10, release=None, preload=True, poolsize=2):
    print '-' * 10 + "  timeout=%s preload=%s release=%s " % (timeout, preload, release) + '-' * 10
    cp = HTTPConnectionPool('httpbin.org', block=True, maxsize=poolsize, timeout=None)
    group = []
    for i in range(3):
        group.append(spawn(timed_read, cp, i, timeout, release=release, preload_content=preload))
    joinall(group)

    current_size = cp.pool.qsize()
    if current_size != poolsize:
        print "poolsize decreased: %s" % current_size


for timeout, preload in product([10, 4, 0.1], [True, False]):
    main(timeout=timeout, preload=preload)
```

Setting a timeout that way may seem stupid, as it could be set on the connection itself, but it's just an example for having a timeout running on a longer running task including several calls into the connection pool. What happens is, that the connections are not properly returned back into the queue, and the queue finally runs empty, and any further call to it blocks forever:

<pre>
----------  timeout=10 preload=True release=None ----------
1 opened
1 finished
0 opened
0 finished
2 opened
2 finished
----------  timeout=10 preload=False release=None ----------
0 opened
0 finished
1 opened
1 finished
2 opened
2 finished
----------  timeout=4 preload=True release=None ----------
1 opened
1 finished
0 timed out
2 timed out
----------  timeout=4 preload=False release=None ----------
1 opened
1 finished
0 opened
0 finished
2 timed out
poolsize decreased: 1
----------  timeout=0.1 preload=True release=None ----------
0 timed out
1 timed out
2 timed out
----------  timeout=0.1 preload=False release=None ----------
0 timed out
1 timed out
2 timed out
poolsize decreased: 0
</pre>

Simply setting `preload_content` to `True` is not an option, as e.g. `False` is set by default by requests.

The main problem is in https://github.com/shazow/urllib3/blob/master/urllib3/response.py#L246 . The `Timeout` does not derive from `Exception`, so `self._original_response` is not closed in the except clause, and therefore the connection is not released in the finally clause.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/31218594-connection-pool-queue-can-still-be-exhausted-by-gevent-timeouts?utm_campaign=plugin&utm_content=tracker%2F192525&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F192525&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Lukasa**
Resolves #805 (again).

This patch ensures that we correctly cleanup connections that fail inside `HTTPConnectionPool.urlopen`, not just ones that fail after opening. This should ensure that weird gevent timeouts during connection setup also correctly eliminate the connection, rather than leaking it.

### CODE REVIEW

File: test/test_connectionpool.py [hunk 1]
```diff
@@ -283,6 +283,28 @@ def test_ca_certs_default_cert_required(self):
             conn = pool._get_conn()
             self.assertEqual(conn.cert_reqs, 'CERT_REQUIRED')
 
+    def test_cleanup_on_extreme_connection_error(self):
+        class RealBad(BaseException):
```
Comment (sigmavirus24) [inline test/test_connectionpool.py]: So, perhaps it's just me, but I'd love to have a comment or docstring for this that describes why we have an exception class inheriting from BaseException and why we're testing that here.

(In reality, I'd love for all tests to have docstrings, but this one especially.)

File: urllib3/connectionpool.py [hunk 2]
```diff
@@ -628,6 +627,12 @@ def urlopen(self, method, url, body=None, headers=None, retries=None,
             err = e
 
         finally:
+            if not clean_exit:
+                # We hit some kind of exception, handled or otherwise. We need
+                # to throw the connection away unless explicitly told not to.
+                conn = conn and conn.close()
```
Comment (sigmavirus24) [inline urllib3/connectionpool.py]: I take it that `conn.close()` will return `None`, right? In other words, the result of this will be a state where the locals has `conn = None` and then `self._put_conn` will add `None` back to the pool?

Then we'll go onto the retry logic around L646?

**Comment by sigmavirus24** [comment]
tl;dr this looks fine except I'd love to get the @ml31415's feedback on this + better commenting around this code so we know what it is and why it is later on when we've all forgotten the joyride that #805 has become. ;)

File: urllib3/connectionpool.py (hunk 2 above)
Comment (Lukasa) [inline urllib3/connectionpool.py]: Exactly right. This pattern was present in the other blocks of code that I consolidated here.

**Comment by Lukasa** [comment]
@sigmavirus24 I think I've added some clarifying material, let me know how you feel about it.

**Comment by shazow** [comment]
This is one of those refactors where I'm really glad we have great coverage for that code. :) 

I suspect this will be an intermediary step towards a further refactor in the future but +1 from me!

**Comment by sigmavirus24** [comment]
:+1: from me as well. Thanks for the comments @Lukasa. I think future us (or the poor souls who replace us) will thank us.

**Comment by Lukasa** [comment]
Yeah, code coverage here was huge, I felt much more comfortable with this change than I would have without it.

I'm going to let someone else merge. =)

**Comment by shazow** [comment]
:custard:
